src/pending.js

```diff
@@ -0,0 +1,128 @@
+var _ = require('lodash');
+var debug = require('debug')('stats-collector:pending');
+var taskcluster = require('taskcluster-client');
+
+const FLUSH_PERIOD = 30; // seconds
+
+// check for pending tasks that aren't really pending with this frequency; should
+// be smaller than MIN_CHECK_AGE
+const CHECK_INTERVAL = 120; // seconds
+
+// only check tasks pending at least this long; this should be lower than any alert
+// thresholds for pending tasks.
+const MIN_CHECK_AGE = 300; // seconds
+
+// calculate the taskKey and scheduled date with the earliest scheduled time for
+// the given workerType; if there are none, returns null and the current time
+var earliest = (state, workerType, now) => {
+  return _.reduce(state.pendingTasks[workerType], (res, scheduled, taskKey) => {
```

Comment:
Oh, I see `state.pendingTasks[workerType]` is actually an object...

I suggest you just make a single list... nested maps will only give you pain.. it's not like any of this will be CPU bound anyways (I suspect). If you wanted to make it faster by using a hashmap I doubt `{}` is the way to go, you would probably have to find a real hash-table implementation... I bet `{}` iterates over the keys under the hood.

To be clear suggest you use a data model: `state.pendingTasks = [ {workerType: ..., scheduled, taskIdRunId: ..} ]`

Then it just becomes `_.maxBy(_.filter(state.pendingTasks, {workerType}), 'scheduled');`